### PR TITLE
Stm-113-Filtering

### PR DESCRIPTION
**Pull request by idzidz**
# General

image

New directory added:

- Contains "data" folder stolen from api branch w/ nasdaq.csv and tsx.csv
- Created stock_search.py function that returns the first match of a stock given either a stock symbol or name. If given "tsla" (ticker) it will find and store "TSLA": "Tesla Inc." as a dict. Or the other way around, if given "tesla" (name) it will still find and store "TSLA": "Tesla Inc"
- I kept it to first match because the Twitter scraper uses these results as its search terms. Partial matches would result in potentially hundreds of search queries that the Twitter scraper would have to go through (and maybe never finish).

# Twitter

image

- Twitter scraper will use cli args as the search & filter params if they are provided. If not, original functionality remains. 
- Filtering bodies of tweets seems to be a bit redundant as all my test cases resulted in 100% of searched posts containing the hashtag in the tweet body, such as #TSLA.
- File output will be tweets that only contain the searched stock. As previously mentioned, this seems to just be all of them.
- Used basic cli arg inputs separated by spaces rather than argparse w/ "--stocks x y z" as it doesn't seem required (yet) and I was lazy.

# Reddit 

image

- Reddit scraper will use cli args as the filter params if they are provided. If not, original functionality remains.
- Will still use provided subreddits to scan which can be found in the CrawlScrape/reddit/reddit_resource/subreddits.py. Has been changed to just look at tesla stuff for now
- If arguments are provided, it will output the original unfiltered file, along with the filtered posts in the appropriate directory based on the stock it filtered for. 
image

# The End

Lots of future TODOs and fixes but should be useful for demo. Have not tested on MacOS.

### CODE REVIEW

File: twitter/selenium_twitter_scraper.py [hunk 1]
```diff
@@ -241,7 +260,8 @@ def store_tweet_data_payload(searchQuery, dataPayload):
 driver = Chrome(service=Service(ChromeDriverManager().install()), options=options) #Firefox: "Firefox" | Edge: "options = EdgeOptions(); options.use_chromium = True; driver = Edge(options=options)"
 logger.debug("--- Created Chrome driver ---  ")
 
-driver.implicitly_wait(20) # Better Practice to use this than time.sleep() (Unlike 'time.sleep()', 'driver.implicitly_wait()' is NOT a FIXED wait time) [Gives more time to load webpage to find elements]
```
Comment (Marco-Cen) [inline twitter/selenium_twitter_scraper.py]: Bring this back! From my experience, should reduce scroll margin error received

Comment (idzidz) [reply twitter/selenium_twitter_scraper.py]: yes daddy

File: twitter/selenium_twitter_scraper.py [hunk 2]
```diff
@@ -208,22 +222,27 @@ def store_tweet_data_payload(searchQuery, dataPayload):
...
 # (In order to run CI agent account on Ubuntu server (within Jenkins Build Environment)
-options.add_argument("--no-sandbox") # Bypass OS security model (Has to be first option) 
+options.add_argument("--no-sandbox") # Bypass OS security model (Has to be first option)
 options.add_argument("--disable-dev-shm-usage") # overcome limited resource problems
 options.add_argument(f'user-agent={user_agent}') # Needed for headless mode
-# options.add_argument('--headless') # Runs Chrome Driver without actual browser [NOTE: Comment out to debug WITH browser]
+options.add_argument('--headless') # Runs Chrome Driver without actual browser [NOTE: Comment out to debug WITH browser]
+
+# These arguments fix the error of headless not launching fast enough, resulting in the "cannot find element" error.
+options.add_argument("--proxy-server='direct://'")
```
Comment (Marco-Cen) [inline twitter/selenium_twitter_scraper.py:243]: These flag arguments are present in below commented lines of file FYI

**Review by Marco-Cen — CHANGES_REQUESTED** [review]
- Minor code changes in Twitter: Reuse existing code; return to same value of implicit wait
- Data collection: Upon running REDDIT scraper. Only 1 of 2 argument files searched and saved?

**Review by Marco-Cen — APPROVED** [review]
LGTM

**Pull request merged**